### PR TITLE
Modernize

### PR DESCRIPTION
**Pull request by steveoh**
This change updates some of our older tools

- pylint -> ruff
- yapf -> black
- Update the ci pipeline to use release please

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -4,66 +4,61 @@
...
-            'pylint~=2.11',
-            'pytest-cov~=4.0',
-            'pytest-instafail~=0.4',
-            'pytest-isort>=2,<4',
-            'pytest-pylint~=0.18',
-            'pytest-watch~=4.2',
-            'pytest>=6,<8',
-            'yapf~=0.31',
+        "tests": [
+            "pytest-cov==4.*",
```
Comment (stdavis) [inline setup.py:46]: Did you want to switch to some other standard for specifying dependency versions or did we decide that it doesn't matter?

File: src/projectname/version.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
-"""A single source of truth for the version in a programatically-accessible variable.
+"""A single source of truth for the version in a programmatically-accessible variable.
 
 This must only include a single line: __version__ = 'x.y.z'
 """
 
-__version__ = '1.0.0'
+__version__ = "1.0.0"
```
Comment (stdavis) [inline src/projectname/version.py:6]: I'm guessing that this won't work with release please without some additional config. Maybe the extra file param?

**Review by stdavis — CHANGES_REQUESTED** [review]

File: src/projectname/version.py (hunk 2 above)
Comment (stdavis) [reply src/projectname/version.py:6]: Nevermind! I guess that it works! https://github.com/agrc/python/pull/32/files

**Review by stdavis — APPROVED** [review]

File: setup.py (hunk 1 above)
Comment (steveoh) [reply setup.py:46]: it didn't matter

File: src/projectname/version.py (hunk 2 above)
Comment (steveoh) [reply src/projectname/version.py:6]: yeah i never was a big fan of this. how do we test it? @jacobdadams

Comment (stdavis) [reply src/projectname/version.py:6]: I like the idea of having a single source of truth for the version number that is accessible from the code. But maybe it's not as big of a deal now that we auto-bump. What's an alternative?

Comment (jacobdadams) [reply src/projectname/version.py:6]: Yeah, this is an attempt to make the version number accessible by code for logging/reporting purposes (ie, if we get a status email we can know if it's using the most recent version or not). It was added in a world where versions were manually controlled, so auto-bumping definitely changes things.

**Review by jacobdadams — CHANGES_REQUESTED** [review]
Why are we changing all the `'` to `"`? is that a default in black that needs to be updated?

**Comment by jacobdadams** [comment]
Thinking about this a little more, I'm not sold on the changes from pylint and yapf to ruff and black. Right now it feels like "chasing the new shiny" by you guys who (to me) seem to live mostly in the javascript world that would impact myself and others in the office that I'm trying to teach/convince to use this template. 

I'd like to see a little more explanation and discussion before introducing a change that will either require updating a lot of existing projects or require juggling two different sets of build tools for new and old projects.

**Comment by steveoh** [comment]
> Why are we changing all the `'` to `"`? is that a default in black that needs to be updated?

Black is [opinionated](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#strings). You can read the [FAQ](https://black.readthedocs.io/en/stable/faq.html) for some answers to those types of questions. 

> Right now it feels like "chasing the new shiny"

I'm sorry it feels like we are chasing new and shiny but I have to disagree with you. This would be the first python tool chain change you have experienced since working for UGRC. Communities evolve and so do tools and the support they are receiving. We like to choose tools that are well supported and adopted by the community. 

Black has been around for ~5 years and the python community has adopted it with open arms. YAPF just started receiving updates again but was mostly abandoned for the last few years. Ruff is newer. But its performance numbers and the fact that it replaces isort should be enough reason for all of us to rejoice in the change. The shift from configurable formatting to opinionated/uncompromising has happened in multiple programming languages/communities and will be the path going forward for some time. It's in our best interest to follow these trends.

> ... you guys who (to me) seem to live mostly in the javascript world that would impact myself and others in the office that I'm trying to teach/convince to use this template.

Scott and I do not live only in the javascript world. We write code in all the languages and not that it matters but probably more so than most of the others in the office combined. The impact of this change affects us all and should be mostly transparent. It should not affect the adoption of this template in anyway. New project will have different tools that on the surface perform the same job. I would be surprised if anyone in the office is aware of or cares what is formatting or linting their code. The vscode extension suggestions and settings.json will configure their environment to just work™ no matter what project they are working on. People can choose to migrate their projects if they want to use the improved tooling. There is no mandate or requirement to migrate tools. Build tooling does not change. Black also has [documentation](https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html) how how to not ruin your git blame with a style change.

**Comment by stdavis** [comment]
I feel your pain, Jake. I'm not one that likes to chase after the new shiny stuff. It's hard for me to wrap my head around new tools and sometimes I push back. I think that's a good thing. I've also learned that Steve does not get offended by these types of questions and neither do I.

This change seems to be a good one to me. I'm very high on more opinionated, less configurable tools. Especially when it comes to code formatting and linting. I learned long ago that it's not worth arguing over things like `'` vs. `"` or where to add indents. I feel like JavaScript lost years of its life arguing over things like this years ago. Then [prettier](https://prettier.io/) came along and everyone got back to actually building stuff. Life gets SO much easier when you just go with whatever the prevailing opinionated tool is and move forward.

I've tried black and it's really a painless switch especially if you take the time to [apply the format commit the right way](https://black.readthedocs.io/en/stable/guides/introducing_black_to_your_project.html). Once I made the switch, I immediately noticed the performance improvement and then got back to not worrying about how my code was formatted.

I have less experience with ruff, but the fact that it seems to have huge momentum and replaces isort is enough of a reason for me to switch and see how it goes. I've never been able to get isort to work consistently between my machine and CI.

For others in the office using this template, I can't see how this would affect them much at all. The template makes it very easy to get a project going without having to worry about what tools it's using. Just bootstrap the project, install the dependencies, and get to work. Existing projects can be brought into alignment as you work on them. I don't feel like it's a huge push to upgrade everything right away.

**Comment by stdavis** [comment]
See the first article in [this newsletter](https://bytes.dev/archives/205?ck_subscriber_id=1219737903) for a much better (and more entertaining) overview than I gave of how prettier affected the JavaScript ecosystem.

**Pull request merged**